[Parser] Don't modify the current token kind when cutting off parsing

ahoppen opened the pull request:
Instead, let the lexer know that it should interpret the next token to be lexed as an EOF token. This makes parsing cutoff a lot more stable because the parser is not expecting the current token kind to change in a lot of places.

Technically, this causes two more tokens to be parsed than in the previous implementation (namely `Parser.Tok` and `Lexer.NextToken`) but since we bail out at a fairly low nesting level, this shouldn't make a big difference.

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on validation-test/compiler_crashers_2_fixed/parser-cutoff.swift:
```diff
@@ -0,0 +1,793 @@
+if a > 1 {
+if a > 1 {
+if a > 1 {
+    
```
I assume this was from me, obviously doesn't matter but could remove if you wanted it to be a little neater :P

ahoppen replied:
Whatever. The file hit the issue before and now doesn’t hit it any longer so I don’t want to change it anymore.

bnbarham reviewed (approved):
Nice!

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[Parser] Don't modify the current token kind when cutting off parsing

Previously, when we reached the maximum nesting level, we changed the current token’s kind to an EOF token. A lot of places in the parser are not set up to expect this token change. The intended workaround was to check whether pushing a structure marker failed (which would change the token kind) and bail out parsing if this happened. This was fragile and caused assertion failures in assert builds.

Instead of changing the current token’s kind, and failing to push the structure marker, let the lexer know that it should cut off lexing, essentially making the input buffer stop at the current position. The parser will continue to consume its current token (`Parser.Tok`) and the next token that’s already lexed in the lexer (`Lexer.NextToken`) before reaching the emulated EOF token. Thus two more tokens are parsed than before, but that shouldn’t make much of a difference.

ahoppen commented:
I refactored the PR a bit, reverting a fair bit of https://github.com/apple/swift/pull/19631 because the checks for `StructureMarkerRAII::isFailed` are no longer necessary with the new cut-off model.

@bnbarham Could you take another look at the changes?

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on validation-test/compiler_crashers_2_fixed/parser-cutoff.swift:
```diff
@@ -0,0 +1,791 @@
+if a > 1 {
+if a > 1 {
+
+// RUN: not %swift-frontend -typecheck %s
```
The crash here was the assertion failure right? Would it be worth adding `// REQUIRES: asserts` or a comment to that effect in here? The test does work either way so maybe doesn't matter, but IMO it'd be nice to differentiate it from the existing `structure_overflow*.swift` ones.

ahoppen replied:
Added a comment.

bnbarham reviewed (approved):
Nice, much cleaner!

ahoppen commented on lib/Parse/Parser.cpp:
```diff
@@ -405,6 +405,9 @@ namespace {
   ASTContext &Ctx;
+  /// The lexer that is being used to lex the source file. Used to query whether
+  /// lexing has been cut off.
+  Lexer *BaseLexer;
```
Good idea. Changed it to a reference.

ahoppen commented:
@swift-ci Please smoke test
